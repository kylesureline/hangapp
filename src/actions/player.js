import database from '../firebase/firebase';

export const initializeGame = (word, guessesRemaining) => ({
  type: 'INIT_GAME',
  answer: word,
  guessesRemaining
});

export const guessLetter = (letter) => ({
  type: 'GUESS_LETTER',
  letter
});

export const updateGuessedWord = (wordArray) => ({
  type: 'UPDATE_GUESSED_WORD',
  wordArray
});

export const wrongGuess = () => ({
  type: 'WRONG_GUESS'
});

export const setPlayer = (player) => ({
  type: 'SET_PLAYER',
  player
});

export const setPlayerStats = (stats) => ({
  type: 'SET_PLAYER_STATS',
  stats
});

export const setPastGames = (pastGames) => ({
  type: 'SET_PAST_GAMES',
  pastGames
});

export const startSetPlayer = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`players/${uid}`)
      .once('value')
      .then((snapshot) => {
        const pastGames = [];

        for(let game in snapshot.val().pastGames) {
          pastGames.push(snapshot.val().pastGames[game]);
        }

        const stats = snapshot.val().stats;

        dispatch(setPlayerStats(stats));
        dispatch(setPastGames(pastGames));
      });

  };
};
