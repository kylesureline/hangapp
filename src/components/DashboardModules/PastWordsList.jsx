import React from 'react';
import { connect } from 'react-redux';
import PastWordsListItem from './PastWordsListItem.jsx';

export const PastWordsList = (props) => (
  <div className="past-games">
    <div className="list-header">
      <div>Word</div>
      <div className="show-for-desktop">Date</div>
    </div>
    {
      props.pastGames.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No past games available</span>
        </div>
      ) : (
        props.pastGames.sort((a, b) => {
          return b.playedAt - a.playedAt;
        }).map((game, index) => {
          if(index < 10) {
            return <PastWordsListItem key={game.playedAt} {...game} />
          }
        })
      )
    }
  </div>
);

const mapStateToProps = (state) => ({
  pastGames: state.player.pastGames || []
});

export default connect(mapStateToProps)(PastWordsList);
