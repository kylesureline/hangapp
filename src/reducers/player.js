const defaultGameState = {
  answer: {
    word: '',
    type: '',
    def: ''
  },
  guessedLetters: [],
  guessedWord: [],
  guessesRemaining: 0,
  stats: {
    wins: 0,
    losses: 0
  }
};

export default (state = defaultGameState, action) => {
  switch(action.type) {
    case 'GUESS_LETTER':
      return {
        ...state,
        guessedLetters: [...state.guessedLetters, action.letter]
      };
    case 'UPDATE_GUESSED_WORD':
      return {
        ...state
      };
    case 'WRONG_GUESS':
      return {
        ...state,
        guessesRemaining: state.guessesRemaining -= 1
      };
    case 'SET_PLAYER_STATS':
      return {
        ...state,
        stats: action.stats
      };
    case 'SET_PAST_GAMES':
      return {
        ...state,
        pastGames: action.pastGames
      };
    case 'SET_GUESSES_REMAINING':
      return {
        ...state,
        guessesRemaining: action.guessesRemaining
      }
    case 'CHOOSE_DEFAULT_WORD':
      return {
        ...state,
        answer: {
          word: 'hangapp',
          type: 'noun',
          def: 'A hangman game built as a React PWA by Kyle Scheuerlein'
        }
      };
    case 'CHOOSE_RANDOM_WORD':
      return {
        ...state,
        answer: action.answer
      }
    case 'ADD_WORD':
      return {
        ...state,
        pastGames: [
          ...state.pastGames,
          action.answer
        ]
      }
    case 'ADD_WIN':
      return {
        ...state,
        stats: {
          ...state.stats,
          wins: state.stats.wins + 1
        }
      }
    case 'ADD_LOSS':
      return {
        ...state,
        stats: {
          ...state.stats,
          losses: state.stats.losses + 1
        }
      }
    default:
      return state;
  }
};
