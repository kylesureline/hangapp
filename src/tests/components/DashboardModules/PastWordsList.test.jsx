import React from 'react';
import { shallow } from 'enzyme';
import { PastWordsList } from '../../../components/DashboardModules/PastWordsList.jsx';
import { player } from '../../fixtures/player';

test('should render PastWordsList with no past games correctly', () => {
  const wrapper = shallow(<PastWordsList pastGames={[]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render PastWordsList with past games correctly', () => {
  const wrapper = shallow(<PastWordsList pastGames={player.pastGames} />);
  expect(wrapper).toMatchSnapshot();
});
