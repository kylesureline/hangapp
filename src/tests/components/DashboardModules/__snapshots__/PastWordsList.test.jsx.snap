// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PastWordsList with no past games correctly 1`] = `
<div
  className="past-games"
>
  <div
    className="list-header"
  >
    <div>
      Word
    </div>
    <div
      className="show-for-desktop"
    >
      Date
    </div>
  </div>
  <div
    className="list-item list-item--message"
  >
    <span>
      No past games available
    </span>
  </div>
</div>
`;

exports[`should render PastWordsList with past games correctly 1`] = `
<div
  className="past-games"
>
  <div
    className="list-header"
  >
    <div>
      Word
    </div>
    <div
      className="show-for-desktop"
    >
      Date
    </div>
  </div>
  <PastWordsListItem
    def="in this"
    id="15"
    key="herein"
    playedAt={1565717060602}
    type="adverb"
    won={false}
    word="herein"
  />
  <PastWordsListItem
    def="instruction especially of an individual"
    id="14"
    key="tutelage"
    playedAt={1565717033093}
    type="noun"
    won={false}
    word="tutelage"
  />
  <PastWordsListItem
    def="to make invalid; especially : to weaken or destroy the cogency of"
    id="13"
    key="invalidate"
    playedAt={1565717015798}
    type="verb"
    won={false}
    word="invalidate"
  />
  <PastWordsListItem
    def="elderberry"
    id="12"
    key="elders"
    playedAt={1565716947670}
    type="noun"
    won={false}
    word="elders"
  />
  <PastWordsListItem
    def="the smallest of three dimensions"
    id="11"
    key="thickness"
    playedAt={1565716854400}
    type="noun"
    won={false}
    word="thickness"
  />
  <PastWordsListItem
    def="grandfather"
    id="10"
    key="grandad"
    playedAt={1565716749582}
    type="noun"
    won={false}
    word="grandad"
  />
  <PastWordsListItem
    def="to mix again"
    id="9"
    key="remixes"
    playedAt={1565716729632}
    type="verb"
    won={false}
    word="remixes"
  />
  <PastWordsListItem
    def="the value set in Anglo-Saxon and Germanic law upon human life in accordance with rank and paid as compensation to the kindred or lord of a slain person"
    id="8"
    key="wergild"
    playedAt={1565716708944}
    type="noun"
    won={false}
    word="wergild"
  />
  <PastWordsListItem
    def="an ability to keep from getting upset or offended by the things other people say and do"
    id="7"
    key="thickskin"
    playedAt={1565716696718}
    type="noun"
    won={false}
    word="thickskin"
  />
  <PastWordsListItem
    def="inlaid decoration of tortoiseshell, yellow metal, and white metal in cabinetwork"
    id="6"
    key="buhls"
    playedAt={1565716667188}
    type="noun"
    won={false}
    word="buhls"
  />
</div>
`;
